File: bowling/Bowling04/Bowling04/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Bowling04
{
    class Game
    {
        private readonly int[] _rolls = new int[21];
        private int _rollIndex = 0;

        internal void Roll(int pinsDown)
        {
            _rolls[_rollIndex++] = pinsDown;
        }

        internal int GetScore()
        {
            int frameIndex = 0;
            int sum = 0;

            for (int i = 0; i < 10; i++)
            {
                if (FrameIsStrike(frameIndex))
                {
                    sum += SumStrikeFrame(frameIndex);
                    frameIndex++;
                }
                else if (FrameIsSpare(frameIndex))
                {
                    sum += SumSpareBonus(frameIndex);
                    frameIndex += 2;
                }
                else
                {
                    sum += SumNormalFrame(frameIndex);
                    frameIndex += 2;
                }
            }

            return sum;   
        }

        private bool FrameIsStrike(int frameIndex)
        {
            return _rolls[frameIndex] == 10;
        }

        private int SumStrikeFrame(int frameIndex)
        {
            return 10 + _rolls[frameIndex + 1] + _rolls[frameIndex + 2];
        }

        private int SumNormalFrame(int frameIndex)
        {
            return _rolls[frameIndex] + _rolls[frameIndex + 1];
        }

        private int SumSpareBonus(int frameIndex)
        {
            return 10 + _rolls[frameIndex + 2];
        }

        private bool FrameIsSpare(int frameIndex)
        {
            return _rolls[frameIndex] + _rolls[frameIndex + 1] == 10;
        }
    }
}

File: bowling/Bowling20/src/Game.java

public class Game {

    int[] rolls = new int[21];
    int rollIndex = 0;

    public void roll(int pinsDown) {
        rolls[rollIndex++] = pinsDown;
    }

    public int getScore() {

        int score = 0;
        int rollCounter = 0;

        for (int frameIndex = 0; frameIndex < 10; frameIndex++) {

            if(isStrike(rollCounter)) {
                score += scoreStrike(rollCounter);
                rollCounter += 1;
            }
            else if(isSpare(rollCounter)) {
                score += scoreSpareFrame(rollCounter);
                rollCounter += 2;
            } else {
                score += scoreStandardFrame(rollCounter);
                rollCounter += 2;
            }
        }

        return score;
    }

    private boolean isStrike(int rollCounter) {

        return rolls[rollCounter] == 10;
    }

    private boolean isSpare(int rollCounter) {

        return rolls[rollCounter] + rolls[rollCounter + 1] == 10;
    }

    private int scoreStrike(int rollCounter) {

        return 10 + rolls[rollCounter + 1] + rolls[rollCounter + 2];
    }

    private int scoreSpareFrame(int rollCounter) {

        return 10 + rolls[rollCounter+1];
    }

    private int scoreStandardFrame(int rollCounter) {

        return rolls[rollCounter] + rolls[rollCounter+1];
    }
}
